### ISSUE TITLE
langtags.json validation errors

### ISSUE THREAD

**Issue by mcdurdin** [silnrsi/langtags#5]
There are a few areas where langtags.json 1.1.1 does not validate.

## `"iana"` property type

Most important is the property `"iana"` is documented as type `"string"`. However, entry `"add-Latn-CM"` has the following content (and 425 other entries are the same):

```
  iana: [ 'Lidzonka', 'Dzodinka' ],
```

It would be better for it to always be an array -- mixing strings and arrays is just a special case irritation for parsing.

## additional undocumented property `"rod"`

24 entries include an undocumented property `"rod"`:

```
{
  full: 'abt-Latn-PG-x-maprik',
  iana: 'Ambulas - Maprik',
  iso639_3: 'abt',
  localnames: [ 'Ambulas' ],
  name: 'Ambulas',
  names: [ 'Abelam', 'Abulas', 'Ambulas - Maprik' ],
  region: 'PG',
  regionname: 'Papua New Guinea',
  rod: '00444',
  script: 'Latn',
  sldr: false,
  tag: 'abt-x-maprik',
  tags: [ 'abt-Latn-x-maprik', 'abt-PG-x-maprik' ],
  windows: 'abt-Latn-x-maprik'
}
```

## Invalid tags

There are also a number of invalid `"tags"` and `"fulltags"`. Most of these have over-long extension subtags (which should have a maximum of 8 characters), but one is just an invalid bcp 47 tags -- out of order -- and none are listed as grandfathered tags. The following regexes are full matches:

```
brv-(Thai|TH)-x-(dongluang|khongchiem|sakonnakon)
cek-(Latn-)?(MM-)?x-asangkhongso
cek-(Latn-)?(MM-)?x-khawngtuu
dao-(Latn-)?(MM-)?x-khengdaai
dgl-(Copt-)?(SD-)?x-oldnubian
ers-(Zzzz-)?(CN-)?x-ersushaba
fia-(Copt-)?(SD-)?x-oldnubian
mnc-(Mong-)?(CN-)?x-oldmanchu
nst-(Latn-)?(MM-)?x-moshanghawa
onw-(Copt-)?(SD-)?x-oldnubian
sgn-(Zxxx-)?MY-(Zxxx-)?MM
sgn-MY-Zxxx
sgn-Zxxx-MY-mm
tew-(Latn-)?(US-)?x-santaclara
tzo-(Latn-)?(MX-)?x-sanandres
tzo-(Latn-)?(MX-)?x-zinacantan
xnz-(Copt-)?(EG-)?x-oldnubian
```

I have created a [JSON schema](https://gist.github.com/mcdurdin/eff65bb2d77034525120d48cbe280d30) that includes these exceptions and hence validates. However, either the documentation should be updated to match the schema, or the issues should be addressed and the schema tightened.

**Comment by mhosken** [comment]
Three main issues here:

1. iana. I need input from folks here. This was this way in 1.1.0 and that is in production, so I'm guessing nobody is actually using the "iana" field otherwise they would have already complained.
2. Easy enough to add to the document
3. Sigh. I could have sworn I was allowed any number of characters after x- and that variants were 5-11. I'll work to shrink them. But for now I suggest you stretch your schema to allow any number of characters after x- The biggest offender is asangkhongso at 12. I don't think we need to grandfather these in.
4. sgn-MY-Zxxx-MM. This was designed using extlang, but breaks its rules. We should change this it sgn-Zxxx-MM and just go with sign language as used in Myanmar. This would probably solve the Windows field problem as well.

I do hope you will share you schema and perhaps a test for it.

**Comment by mcdurdin** [comment]
> I do hope you will share you schema and perhaps a test for it.

The JSON schema is linked above (and [here](https://gist.github.com/mcdurdin/eff65bb2d77034525120d48cbe280d30) again too).

I'd write a test but I don't do much python so it'd take me longer than I really have at present. It should be straightforward to apply a [schema validator](https://json-schema.org/implementations.html#validator-python).

Node.js junk test that I used:
```js
const fs = require('fs');
const Ajv = require('ajv');

const json = JSON.parse(fs.readFileSync('langtags.json'));

var ajv = new Ajv();
var validate = ajv.compile(JSON.parse(fs.readFileSync('langtags.schema.json')));
var valid = validate(json);

if(!valid) {
  validate.errors.forEach(e => {
    console.log(e);
    if(e.dataPath)
      console.log(eval('json'+ e.dataPath));
    console.log('');
  });
}
```

**Comment by darcywong00** [comment]
Just noting some recent validation errors as of Mar 2022:

```json
{
    "params": { "additionalProperty" : "'unwritten'" },
    "message": "'must NOT have additional properties'"
},
{
    "params": { "missingProperty": "'variants'" },
    "message": "must have required property 'variants'"
}

**Comment by darcywong00** [comment]
I think 1.1.1 validated.
But 1.1.2 added `_conformance` so that probably needs to get added to the schema too.

Related to keymanapp/api.keyman.com#176

Can possibly update schema from sillsdev/SpeechAnalyzer#103

**Comment by darcywong00** [comment]
I believe @mhosken resolved this in 27203732e46d130b40b94e9078e99bd065291197

**Comment by mcdurdin** [comment]
@mhosken are these resolved?
